src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +121,287 @@ object RuleUtils {
+
+        // Clear sortColumnNames as BucketUnion does not keep the sort order within a bucket.
+        val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
+
+        // Merge index plan & newly shuffled plan by using bucket-aware union.
+        BucketUnion(
+          Seq(indexPlan, transformPlanToShuffleUsingIndexSpec(bucketSpec, planForAppended)),
+          bucketSpec)
+      } else {
+        // If bucketing is not necessary (e.g. FilterIndexRule), we use [[Union]] to merge
+        // the appended data without additional shuffle.
+        Union(indexPlan, planForAppended)
+      }
+    } else {
+      indexPlan
+    }
+  }
+
+  /**
+   * Transform the current plan to read the given appended source files.
+   *
+   * The result will be merged with the plan which is reading index data
+   * by using [[BucketUnion]] or [[Union]].
+   *
+   * @param spark Spark session.
+   * @param indexSchema Index schema used for the output.
+   * @param originalPlan Original plan.
+   * @param filesAppended Appended files to the source relation.
+   * @return Transformed linear logical plan for appended files.
+   */
+  private def transformPlanToReadFromAppendedFiles(
+      spark: SparkSession,
+      indexSchema: StructType,
+      originalPlan: LogicalPlan,
+      filesAppended: Seq[Path]): LogicalPlan = {
+    // Replace the location of LogicalRelation with appended files.
+    val planForAppended = originalPlan transformDown {
+      case baseRelation @ LogicalRelation(fsRelation: HadoopFsRelation, baseOutput, _, _) =>
+        // Set the same output schema with the index plan to merge them using BucketUnion.
+        val updatedOutput =
+          baseOutput.filter(attr => indexSchema.fieldNames.contains(attr.name))
+        val newLocation =
+          new InMemoryFileIndex(spark, filesAppended, Map(), None)
+        val newRelation =
+          fsRelation.copy(
+            location = newLocation,
+            dataSchema = StructType(indexSchema.filter(baseRelation.schema.contains(_))))(spark)
+        baseRelation.copy(relation = newRelation, output = updatedOutput)
+    }
+    assert(!originalPlan.equals(planForAppended))
+    planForAppended
+  }
+
+  /**
+   * Transform the plan to perform on-the-fly Shuffle the data based on bucketSpec.
+   *
+   * Pre-requisite
+   * - The plan should be linear and include 1 LogicalRelation.
+   *
+   * @param bucketSpec Bucket specification used for Shuffle.
+   * @param plan Plan to be shuffled.
+   * @return Transformed plan by injecting on-the-fly shuffle with given bucket specification.
+   */
+  private[rules] def transformPlanToShuffleUsingIndexSpec(
+      bucketSpec: BucketSpec,
+      plan: LogicalPlan): LogicalPlan = {
+    // Extract top level plan including all required columns for shuffle in its output.
+    // This is located inside this function because of bucketSpec.bucketColumnNames.
```

Comment:
You can pass a tuple to `unapply` and update `transformDown` below accordingly, so this is not a hard requirement. But now that this function is refactored, it can live here (and maybe we can remove this comment as well).